### ISSUE TITLE
Support for migrations scripts

### ISSUE THREAD

**Issue by bglidwell** [strapi/strapi#8925]
### Please describe your feature request
Add support for migrations scripts in strapi core similar to how 3.4.0 migration was handled

- [x] **I have created my request on the Product Board before I submitted this issue**
- [x] **I have looked at all the other requests on the Product Board before I submitted this issue**

### Summary

Rather than trying to execute the migration script for 3.4.0 on all my servers, I instead created a migration.js function which exported the idempotent function from the migrate-3.4.0 script. 

![image](https://user-images.githubusercontent.com/14822908/102516869-f8c4a700-4054-11eb-890c-bdb932e61e73.png)

migration.js
```
module.exports = () => {
	strapi.log.info('Running Migrations Script')

	const ONE_RELATIONS = ['oneToOne', 'manyToOne', 'oneWay'];
	
	const isSortableRFAssoc = a =>
		ONE_RELATIONS.includes(a.nature) && !['created_by', 'updated_by'].includes(a.alias);
	
	const migrate = async () => {
		const contentTypeService = strapi.plugins['content-manager'].services['content-types'];
	
		for (const uid of Object.keys(strapi.contentTypes)) {
			const modelDef = strapi.getModel(uid);
			const manyRelationFields = modelDef.associations.filter(isSortableRFAssoc);
			if (manyRelationFields.length) {
				const conf = await contentTypeService.findConfiguration({ uid });
				manyRelationFields.forEach(assoc => {
					try {
						conf.metadatas[assoc.alias].list.sortable = true;
					} catch (e) {
						// silence
					}
				});
				await contentTypeService.updateConfiguration({ uid }, conf);
			}
		}
	};
	
	migrate()
		.catch(e => {
			strapi.log.error('Migration Error:', e);
			process.exit(1);
		})
		.then(() => {
			strapi.log.info('Migration Script successfully finished! 🎉');
		});
};
```

This allowed me to simply call strapi.config.functions.migration() in the bootstrap function.

If migration.js was treated as a core strapi function, this could be extended by saving individual version migrations as exported functions in that file. ie. 
```
module.exports['v3.4.0'] = () => {}
module.exports['v3.5.0'] = () => {}
```

In my opinion, it's a good idea to write migrations this way anyway so that anyone upgrading will just be migrated appropriately and the migrations.js script handles the sequence of events that need to follow.

### Why is it needed?
- Simplifies strapi version upgrades, and ensures all migration steps are taken between versions
- Avoids user error during migration process which could lead to loss of data
- Avoids the necessity of manually executing migration scripts. 
- Allows the possibility of "rollback" functions to downgrade changes. 
- Migrations can be written such that modified data is backed up before execution so it can be restored to a safe state in the event of migration failure.
- Promotes thinking of Migrations as idempotent actions that can be checked for

### Suggested solution(s)
- migration.js function in core
- possible method for detecting version mismatch? (ie. does a migration need to run, perhaps holding strapi_version in core store) 
- Monolith or individual version migration scripts 

### Related issue(s)/PR(s)

Let us know if this is related to any issue/pull request.

**Comment by bglidwell** [comment]
I don’t feel like CLI config dump is the solution as it can contain secrets for providers and such. I wouldn’t commit that to a repo or anything, so it either needs to be sanitized and shared or manually restored. Either instance is requiring a pretty manual intervention for any and all CI/CD environments. Treating migrations this way will create friction for docker deployments and other serverless or containerized environments in future.

**Comment by ralphsomeday** [comment]
I mean this is a no brainer and should be added to the 3.5.0 release as for now migrating from v3.3.x to 3.4.0 is just a headache when you have a CI/CD pipeline to automate your full deployment to staging and production. We cannot afford to do manual deployments on highly sensitive environments. Nowadays everything goes via a CI/CD pipeline with automated testing, security and performance testing before releasing to production.

**Comment by bglidwell** [comment]
@ralphsomeday you bring up a good point about automated testing.

**Comment by ralphsomeday** [comment]
I like the proposed solution of @bglidwell the migration should be as silent as possible and at bootstrap so that  when we run the automated tests in our CI/CD pipeline then it is against the latest upgraded version of Strapi. @derrickmehaffy can we put that issue as a high prio on the list of the strapi core team as for now we cannot upgrade to 3.4.0 unless we use @bglidwell current workaround? thanks

**Comment by bglidwell** [comment]
It's worth adding that the solution I proposed still needs a mechanism for running only upon upgrade. An unintended side effect of just sticking it in the bootstrap.js is that it will reset all many-relationship columns to sortable regardless of the custom config you set in the admin list view. 

A solution to this problem I proposed above is perhaps storing the strapi_version in the core config in the database as a variable, such that it can be checked and compared at boot time to run whatever migration script(s) are necessary to bring the schemas up to date.

**Comment by derrickmehaffy** [comment]
> I like the proposed solution of @bglidwell the migration should be as silent as possible and at bootstrap so that  when we run the automated tests in our CI/CD pipeline then it is against the latest upgraded version of Strapi. @derrickmehaffy can we put that issue as a high prio on the list of the strapi core team as for now we cannot upgrade to 3.4.0 unless we use @bglidwell current workaround? thanks

You don't actually need the migration script to update. 

https://strapi.io/documentation/developer-docs/latest/migration-guide/migration-guide-3.3.x-to-3.4.0.html#migrate-configurations

> This migration step is not required. If you prefer doing the configuration updates manually, you can ignore it.

**Comment by ggirodda** [comment]
I suggest to integrate this package, to generalize migrations scripts in strapi:
https://github.com/tj/node-migrate
It could be used also for version update migration scripts. If needed, I can try to integrate it to strapi, maybe creating a plugin. Can someone give me a confirmation that this kind of functionality doesn't yet exist for strapi ? Is there some current development that will integrate something similar ?

**Comment by pelumiojebode** [comment]
New integrations
New formulas

**Comment by ggirodda** [comment]
@pelumiojebode I don't understand your answer, sorry. Can you explain please ?

**Comment by bglidwell** [comment]
@ggirodda That module looks like just what we need. Thanks for pointing it out. I may replace a custom migration implementation in some of my own projects as well.

**Comment by bglidwell** [comment]
May have to modify how it saves migrations that were run to use the core store instead of file system, though.

**Comment by violabg** [comment]
I'm getting this error when trying to migrate

` node migrate-3.4.0.js mc-intra-api/
[2020-12-20T14:54:22.664Z] debug ⛔️ Server wasn't able to start properly.
[2020-12-20T14:54:22.665Z] error error: password authentication failed for user "root"
    at Parser.parseErrorMessage (/var/www/mc-intra-api/node_modules/pg-protocol/dist/parser.js:278:15)
    at Parser.handlePacket (/var/www/mc-intra-api/node_modules/pg-protocol/dist/parser.js:126:29)
    at Parser.parse (/var/www/mc-intra-api/node_modules/pg-protocol/dist/parser.js:39:38)`

**Comment by derrickmehaffy** [comment]
@violabg please create a new issue for bugs, this issue is not the place for that.

**Comment by ggirodda** [comment]
I created this plugin
basically is just node-migrate, with strapi load before the call that requires strapi to be initialized
https://www.npmjs.com/package/strapi-plugin-database-migration
I tested with 3.4.0 script migration, and it works as expected. So I created a migration, using `yarn strapi-migrate-create migrate-3.4.0` and put the script inside (a simplified version)
```javascript
'use strict';

const ONE_RELATIONS = ['oneToOne', 'manyToOne', 'oneWay'];

const isSortableRFAssoc = (a) =>
  ONE_RELATIONS.includes(a.nature) &&
  !['created_by', 'updated_by'].includes(a.alias);

module.exports.up = async function () {
  const contentTypeService =
    strapi.plugins['content-manager'].services['content-types'];

  for (const uid of Object.keys(strapi.contentTypes)) {
    const modelDef = strapi.getModel(uid);
    const manyRelationFields = modelDef.associations.filter(isSortableRFAssoc);
    if (manyRelationFields.length) {
      const conf = await contentTypeService.findConfiguration({ uid });
      manyRelationFields.forEach((assoc) => {
        try {
          conf.metadatas[assoc.alias].list.sortable = true;
        } catch (e) {
          // silence
        }
      });
      await contentTypeService.updateConfiguration({ uid }, conf);
    }
  }
};

module.exports.down = function (next) {
  next();
};
```

then `yarn strapi-migrate-up`

So in my CI I only will put the command `yarn strapi-migrate-up` after the deploy

**Comment by ggirodda** [comment]
For whom using the migration plugin, or wanted to automate the migration to 3.4.4 version (the instagram config update), you can use this (only sql)

```javascript
'use strict';

const _ = require('lodash');

module.exports.up = async function () {
  const knex = strapi.connections.default;
  const result = await knex('core_store').where(
    'key',
    'plugin_users-permissions_grant',
  );

  const value = _.get(result, '0.value') && JSON.parse(result[0].value);

  if (_.get(value, 'instagram')) {
    value.instagram.scope = ['user_profile'];
    await knex('core_store')
      .where('key', 'plugin_users-permissions_grant')
      .update({
        value,
      });
  }
};

module.exports.down = async function () {
  const knex = strapi.connections.default;
  const result = await knex('core_store').where(
    'key',
    'plugin_users-permissions_grant',
  );

  const value = _.get(result, '0.value') && JSON.parse(result[0].value);

  if (_.get(value, 'instagram')) {
    delete value.instagram.scope;
    await knex('core_store')
      .where('key', 'plugin_users-permissions_grant')
      .update({
        value,
      });
  }
};
```

**Comment by SalahAdDin** [comment]
A migration tool is very critical for enterprise and non enterprise users.

**Comment by TheBit** [comment]
https://github.com/strapi/strapi/tree/master/packages/core/database/lib/migrations - v4 undocumented (because untested yet, and down is not implemented yet)

**Comment by derrickmehaffy** [comment]
> > https://github.com/strapi/strapi/tree/master/packages/core/database/lib/migrations - v4 undocumented (because untested yet, and down is not implemented yet)
> 
> And it is very old I guess.

it's not old at all, we just built it in nov/dec

It's undocumented because as @TheBit mentioned it's unstable. You can see an example in this repo I created: https://github.com/derrickmehaffy/strapi-example-v4-custom-db

**Comment by derrickmehaffy** [comment]
Currently already exists, just needs documentation which should be coming shortly (the feature isn't stable yet but we know about it and already plan to work on it).

Going to mark this as closed as technically the script part does exist. See my link above on how to use it for now.
